migrate crml/cennzx-spot to plug 1.0

cowboy-bebug opened the pull request:
Changes:
```
cennzx-spot/Cargo.toml
- uint bumped 0.7.1 -> 0.8 to avoid use of deprecated item warning

cennzx-spot/src/impls.rs
- FeeExchange and BuyFeeAsset bits are commented out
- trait bound `T::AssetId: Into<u64>` added for ExchangeAddressFor trait

cennzx-spot/src/tests.rs
- minor refactoring on build method for ExtBuilder (eg, using `assimilate_storage()`)

cennzx-spot/src/types.rs
- removed `runtime_primitives::traits::As` as it no longer exists in substrate
```

Todo:
- [x] integrate cennzx-spot into runtime module (there are a lot of compile errors)
- [x] add CennzxSpotConfig to chain specs
- [x] fix `div()` method for FeeRate (`As<128>`bit)
- [x] fix TODOs
- [x] check license headers

Closes #60

cowboy-bebug committed:
migrate crml/cennzx-spot to plug 1.0

cennzx-spot/Cargo.toml
- uint bumped 0.7.1 -> 0.8 to avoid use of deprecated item warning

cennzx-spot/src/impls.rs
- FeeExchange and BuyFeeAsset bits are commented out
- trait bound `T::AssetId: Into<u64>` added for ExchangeAddressFor trait

cennzx-spot/src/tests.rs
- minor refactoring on build method for ExtBuilder (eg, using `assimilate_storage()`)

cennzx-spot/src/types.rs
- removed `runtime_primitives::traits::As` as it no longer exists in substrate

note: the crate builds and passes all the tests

cowboy-bebug committed:
redirect dependencies to chore branch

cowboy-bebug committed:
add CennzxSpot and CennzxSpotConfig to runtime and chain spec, respectively

cowboy-bebug committed:
use TryInto<u128> in place for deprecated As<u128> trait

cowboy-bebug commented on crml/cennzx-spot/src/types.rs:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	/// Divide a `TryInto<u128>` supported numeric by a FeeRate
+	pub fn div<N: TryInto<u128>>(lhs: N, rhs: FeeRate) -> rstd::result::Result<u128, &'static str> {
```
`TryInto` is used instead of `Into` as the test for this uses `i32` for lhs:
```rust
assert_ok!(FeeRate::div(10, fee_rate), 9_u128)
```

jordy25519 replied:
Since `safe_div` exists, let's just remove this and it's associated tests. It does the same thing now

cowboy-bebug commented on crml/cennzx-spot/src/types.rs:
```diff
@@ -0,0 +1,158 @@
+	#[test]
+	fn div_works() {
+		let fee_rate = FeeRate::from_percent(110);
+		assert_ok!(FeeRate::div(10, fee_rate), 9_u128); // Float value would be 9.0909
```
Changed to `assert_ok!()` to match fn signature

cowboy-bebug committed:
remove div and related test

cowboy-bebug committed:
run cargo fmt

cowboy-bebug committed:
redirect dependencies to rc1 branch

jordy25519 commented on crml/cennzx-spot/src/impls.rs:
```diff
@@ -0,0 +1,129 @@
+//! Extra CENNZX-Spot traits + implementations
+//!
+use super::Trait;
+// use cennznet_primitives::FeeExchange;
```
Would prefer this commented stuff was removed from this PR (including tests). It can be added back in cleanly in follow ups.

jordy25519 commented on crml/cennzx-spot/src/lib.rs:
```diff
@@ -0,0 +1,858 @@
+};
+use system::{ensure_root, ensure_signed};
+
+// TODO: remove this temp fix for overflow issue when upstream fix ready
```
Looks like this is ok to remove. I think there were a few iterations of "temp fix" and the comments were not updated

jordy25519 commented on crml/cennzx-spot/src/lib.rs:
```diff
@@ -0,0 +1,858 @@
+	type BalanceToU128: From<<Self as generic_asset::Trait>::Balance> + Into<u128>;
+	type U128ToBalance: From<u128> + Into<<Self as generic_asset::Trait>::Balance>;
+
+	//	type AsBalance: From<<Self as generic_asset::Trait>::Balance> + Into<<Self as generic_asset::Trait>::Balance> + As<u128>;
```
This can go also

jordy25519 reviewed (dismissed):
LGTM. Once CI passes and comments addressed

cowboy-bebug committed:
make generic-asset balance configurable

cowboy-bebug committed:
Revert "make generic-asset balance configurable"

This reverts commit 51c28a9e41857f22f7dfcadad32b8a00f7630233.

cowboy-bebug committed:
remove BuyFeeAsset and related tests with some clean up

cowboy-bebug committed:
remove unused imports

cowboy-bebug committed:
add helper methods to initialise generic-asset free balance
- added `fn map_build()` to build a HashMap
- added `fn ext_build()` to build storage from the output of `map_build()`
- added `merge!` macro to merge multiple HashMaps
- edited `with_exchange!` and `with_account!` macro to return tuple
- edited all unit tests accordingly
- added `substrate-state-machine` as a `dev-dependencies`
  because it’s not no_std compatible and only used in tests

cowboy-bebug committed:
change remaining inline merging to use merge! macro

cowboy-bebug committed:
Revert "change remaining inline merging to use merge! macro"

This reverts commit e12de00c266926b9ee5ec0eb2e57c74fa1607b39.

cowboy-bebug committed:
Revert "add helper methods to initialise generic-asset free balance"

This reverts commit 4aa2458b9b88d48f07f1c7bab4a565c4e62928f9.

jordy25519 committed:
Use currency typesfor CENNZX tests (#90)

jordy25519 reviewed (approved):
🚀

aliXsed commented:
@elim0322 As you have just migrated the code and it was hard for me to see which part is your changes to make it work, feel free to disregard any of my reviews that you find not related to the task.

cowboy-bebug commented:
> @elim0322 As you have just migrated the code and it was hard for me to see which part is your changes to make it work, feel free to disregard any of my reviews that you find not related to the task.

Right I'll have a look at the errors and let u know if there's anything
